constellations/orion/src/controllers/classes_by_semesters_controller.go

```diff
@@ -0,0 +1,118 @@
+package controllers
+
+import (
+	"net/http"
+
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/domains"
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/repos"
+	"github.com/gin-gonic/gin"
+)
+
+func GetAllProgramsSemestersClasses(c *gin.Context) {
+	var classSemesterJson domains.ProgramClassesBySemester
+	var listResults []domains.ProgramClassesBySemester
+
+	// Fetch progrmas, semesters, classes from repo functions
+	publishedOnly := ParseParamPublishedOnly(c)
+
+	programs, err := repos.ProgramRepo.SelectAll(publishedOnly)
+	semesters, err := repos.SemesterRepo.SelectAll(publishedOnly)
+	classes, err := repos.ClassRepo.SelectAll(publishedOnly)
+
+	// Convert lists into maps
+	programMap := make(map[string]domains.Program)
+	semesterMap := make(map[string]domains.Semester)
+	classMap := make(map[string]domains.Class)
+
+	for i := 0; i < len(programs); i++ {
+		programMap[programs[i].ProgramId] = programs[i]
+	}
+	for i := 0; i < len(semesters); i++ {
+		semesterMap[semesters[i].SemesterId] = semesters[i]
+	}
+	for i := 0; i < len(classes); i++ {
+		classMap[classes[i].ClassId] = classes[i]
+	}
+
+	// Semester to class list map (maps semesterId to Class object)
+	semesterClassMap := make(map[string][]domains.Class)
+
+	// loop over semesterIds
+	for _, value := range semesterMap {
+		// create Semester Object
+		semesterObj := value
+		var programClassStruct domains.ProgramClass
+		var programObj domains.Program
+
+		// Find classes that have the same semester Id and append to semesterClassMap
+		classSlice := make([]domains.Class, 0)
+		for i := 0; i < len(classes); i++ {
+			if classes[i].SemesterId == semesterObj.SemesterId {
+				classSlice = append(classSlice, classes[i])
+			}
+		}
+		semesterClassMap[semesterObj.SemesterId] = classSlice // list of classes in specific semester
+
+		// create list of programIds in specific semester
+		var programList []string
+		for i := 0; i < len(classSlice); i++ {
+			if Find(programList, classSlice[i].ProgramId) != -1 {
+				continue
```

Comment:
hmm I think it might be better here to have an error! If we have something wrong with our data where things aren't expected, we should error hard - meaning we stop the endpoint and respond with a InternalServerError or something (soft means like a warning / ignore). 

In this case, a class's programId is NOT in the all programs list which is very unusual, so it's something to alert developers.

Reply:
The way I sorted the objects here was to go through a list of classes for each semester and create a list of programs... So basically here, if the programId was already in programList, I just continued the loop onto the next programId. 

I think if there was actually a class's programId that wasn't in the all programs list, it would go wrong where I map from the programId to the full program object at ```programObj = programMap[programList[i]]```. So I guess I'll create an if-then statement there to check if the programId is inside the programMap.